MatrixGame/src/MatrixGameDll.cpp

```diff
@@ -164,166 +67,21 @@ int __stdcall Support() {
-    }
-    catch (...) {
-    }
-
-    try {
-        L3GDeinit();
-    }
-    catch (...) {
-    }
-
-    try {
-        CacheDeinit();
-    }
-    catch (...) {
-    }
-
-#ifdef MEM_SPY_ENABLE
-    try {
-        CMain::BaseDeInit();
-    }
-    catch (...) {
-    }
-#endif
-}
-
 int __stdcall Run(HINSTANCE hinst, HWND hwnd, wchar *map, SRobotsSettings *set, wchar *lang, wchar *txt_start,
                   wchar *txt_win, wchar *txt_loss, wchar *planet, SRobotGameState *rgs) {
-    // try {
-
-    srand((unsigned)time(NULL));
-    SetMaxCameraDistance(set->m_MaxDistance);
-    MatrixGameInit(hinst, hwnd, map, set, lang, txt_start, txt_win, txt_loss, planet);
-    CFormMatrixGame *formgame = HNew(NULL) CFormMatrixGame();
-
-    protectedBlock1(formgame);
-    protectedBlock2(formgame, rgs);
-
-    /*
-    FormChange(formgame);
-
-    timeBeginPeriod(1);
-
-    SETFLAG(g_Flags, GFLAG_APPACTIVE);
-    RESETFLAG(g_Flags, GFLAG_EXITLOOP);
-
-
-    L3GRun();*/
 
-    /*rgs->m_Time=g_MatrixMap->GetPlayerSide()->GetStatValue(STAT_TIME);
-    rgs->m_BuildRobot=g_MatrixMap->GetPlayerSide()->GetStatValue(STAT_ROBOT_BUILD);
-    rgs->m_KillRobot=g_MatrixMap->GetPlayerSide()->GetStatValue(STAT_ROBOT_KILL);
-    rgs->m_BuildTurret=g_MatrixMap->GetPlayerSide()->GetStatValue(STAT_TURRET_BUILD);
-    rgs->m_KillTurret=g_MatrixMap->GetPlayerSide()->GetStatValue(STAT_TURRET_KILL);
-    rgs->m_KillBuilding=g_MatrixMap->GetPlayerSide()->GetStatValue(STAT_BUILDING_KILL);
+    CGame game{};
 
-    timeEndPeriod(1);
+    uint32_t seed = (unsigned)time(NULL);
 
-    MatrixGameDeinit();
+    game.Init(hinst, hwnd, map, seed, set, lang, txt_start, txt_win, txt_loss, planet);
+    
+    CFormMatrixGame formgame;
 
-    FormChange(NULL);
-    HDelete(CFormMatrixGame, formgame, NULL);
-
-    g_Cache->Clear();
-    L3GDeinit();
-    CacheDeinit();
-
-#ifdef MEM_SPY_ENABLE
-    CMain::BaseDeInit();
-#endif
-    */
-    //}
-
-    /*catch(const CException& ex)
-    {
-#ifdef ENABLE_HISTORY
-        CDebugTracer::SaveHistory();
-#endif
-        g_Cache->Clear();
-        L3GDeinit();
-
-        CStr exs(ex.Info());
-        CBuf exb;
-        exb.StrNZ(exs);
-
-        char lpPath[MAX_PATH];
-        strcpy(lpPath,PathToOutputFiles("exception.txt"));
-        size_t size = strlen(lpPath) + 1;
-        wchar_t* wa = new wchar_t[size];
-        mbstowcs(wa,lpPath,size);
-        exb.SaveInFile(wa);//L"exception.txt");
-
-
-        MessageBox(NULL,exs.Get(),"Exception:",MB_OK);
-    }
-    catch(...)
-    {
-        ClipCursor(NULL);
-#ifdef ENABLE_HISTORY
-        MessageBox(NULL,"Unknown bug (history has saved) :(","Exception:",MB_OK);
-        CDebugTracer::SaveHistory();
-#else
-        MessageBox(NULL,"Dont panic! This is just unknown bug happens."
-            "\nCode date: " __DATE__
+    game.RunGameLoop(&formgame);
 
-            ,"Game crashed :(",MB_OK);
-#endif
-    }*/
-    //catch(EXCEPTION_CONTINUE_EXECUTION){}
+    game.SaveResult(rgs);
+    game.SafeFree();
```

Comment:
btw be careful to not ruin the flow here:
before the deleting `CFormMatrixGame` was done after `MatrixGameDeinit` call, but before other stuff clean up. not it's called at the very end, i'm not even sure if it will not crash on the destruction.
please double-check this clean up flow and carefully test it.

Reply:
just a suggestion: you can put `CFormMatrixGame` right into `CGame` class as a member, as turn `SafeFree` function into destructor.
but again - be careful, do not lose `ClipCursor(NULL)` call (from line 86), and test everything.

Reply:
Yes. I checked this moment in game. It works without any problems for me.
I suppose i can do that in next DI refactor